File: src/main/scala/crosscut/CrossCuttingConcerns.scala
package crosscut

import typings.obsidian.mod
import typings.obsidian.mod.FileSystemAdapter
import utils.Utils

import scala.collection.mutable
import scala.language.postfixOps
import scala.scalajs.js
import scala.scalajs.js.Dynamic.literal as l
import scala.scalajs.js.JSConverters.*
import concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.{Failure, Success}

object CrossCuttingConcerns:

  private type DOCNAME   = String
  private type DOCSTRING = String
  private type MARKER    = String
  private type SOLNAME   = String
  private type PATHNAME  = String

  /**
   * ## object CrossCuttingConcerns
   * Setup all the cross cutting concerns for the solution threads. These solution threads are marked by using markers
   * that is picked up from the document md files and placed in the solution folder. ^crosscut-00
   *
   * When applied extract cross cutting text
   */
  def apply(app : mod.App, storyFolder : String,  solutionFolder : String, docFolder: String, markerMapping : String) : Unit =
    //bus
    //bus if a mapping string has been defined then get the mappings : format is 'marker'='mapping-value'
    //
    val fsa = app.vault.adapter.asInstanceOf[FileSystemAdapter]

    val markerMappings : Map[String, String] = if markerMapping.nonEmpty then
      markerMapping
        .split("\n")
        .map( value =>
          if value.contains("=") then
            val lst = value.split("=")
            ( lst(0), lst(1) )
          else
            ( "", "" )
        )
        .toMap
    else
      Map[String, String]()
    //
    // some containers to use later on
    //
    val markerToDocumentMap     = mutable.SortedMap[MARKER, DOCNAME]()
    val markerToStoryMap        = mutable.SortedMap[MARKER, DOCNAME]()
    val documentToMarkerMap     = mutable.HashMap[DOCNAME, List[MARKER]]()
    val storyToMarkerMap        = mutable.HashMap[DOCNAME, List[MARKER]]()
    val solutionToMarkerMap     = mutable.HashMap[SOLNAME, List[MARKER]]()
    val allSolutionFiles        = mutable.HashSet[MARKER]()
    //
    //bus remove all the solution files and the then empty solution folder
    // -----------------------------------------------------------------------------------------------------------------
    val filesToDelete = Utils.listMDFilesInVault(fsa, solutionFolder)

    filesToDelete.foreach(fileName =>
      fsa.remove(fileName).toFuture.foreach(Unit => ())
    )

    fsa.rmdir(solutionFolder, true).toFuture.foreach(Unit => ())
    fsa.mkdir(solutionFolder).toFuture.foreach(Unit => ())

    // -----------------------------------------------------------------------------------------------------------------
    //
    // get all the doc and story files files to scan
    //
    val documentFiles = Utils.listMDFilesInVault(fsa, docFolder)
    val storyFiles = Utils.listMDFilesInVault(fsa, storyFolder)

    //
    // pick up all markers in the doc string doc file by doc file and aggregate the markers
    // before processing them
    //
    val allFutures = mutable.ListBuffer[Future[Unit]]()
    val markerList = mutable.ListBuffer[String]()

    documentFiles.foreach(docFile =>
      allFutures += fsa.read(docFile).toFuture.map(str =>

        val markersMatch = Utils.markerRegExp.findAllMatchIn(str)
        val markersPerDocument = markersMatch.map(marker => str.substring(marker.start, marker.end).trim).toList

        markerList ++= markersPerDocument

        val documentName = docFile.split("/").last

        markersPerDocument.foreach(marker =>
          markerToDocumentMap += (marker -> docFile)
        )

        documentToMarkerMap += (documentName -> markersPerDocument)

      )
    )
    storyFiles.foreach(storyFile =>
      allFutures += fsa.read(storyFile).toFuture.map(str =>

        val markersMatch = Utils.markerRegExp.findAllMatchIn(str)
        val markersPerStory = markersMatch.map(marker => str.substring(marker.start, marker.end).trim).toList

        markerList ++= markersPerStory

        val documentName = storyFile.split("/").last

        markersPerStory.foreach(marker =>
          markerToStoryMap += (marker -> storyFile)
        )

        storyToMarkerMap += (documentName -> markersPerStory)

      )
    )
    //
    // wait for all work to be done
    //
    val waitingForFutures: Future[List[Unit]] = Future.sequence(allFutures.toList)

    waitingForFutures.onComplete {
      case Success(_) =>
        //
        // collect all markers in one list
        // sort them then
        // group by path/name.md excluding the seq number
        //
        val allMarkers = documentToMarkerMap
          .values
          .toList
          .flatten
          .sortWith((s1, s2) =>
            s1 < s2
          )
          .groupBy(by => solutionDocNameFromMarker(solutionFolder, by))
        //
        // write out
        //
        val mdString = StringBuilder()
        allMarkers.foreach((solName, markers) =>
          //
          // build link to story, filter the story markers as well
          //
          val markerToStory = markerToStoryMap.filter((marker, Story) =>
            markers.head.startsWith(marker.split("-").dropRight(1).mkString("-"))
          )
          //
          // setup die story links first
          //
          val mdString = StringBuilder("## Requirement\n")
          markerToStory.foreach((marker, story) =>
             mdString ++= s"""![[${markerToStoryMap(marker)}#${marker}]]\n"""
          )
          mdString ++= "## Solution\n"
          markers.foreach(marker =>
            //
            // build links to document thread
            //
            mdString ++= s"""![[${markerToDocumentMap(marker)}#${marker}]]\n"""
          )
          val marker = markers.head.drop(1).split("-").dropRight(1).mkString("-")
          val solNameWithPath = getSolutionFileName(marker, s"$solName", markerMappings)
          //
          // create the folder path if required and write out text
          //
          Utils.makeDirInVault(fsa, solNameWithPath)
          fsa.write(solNameWithPath, mdString.toString())
        )
      case Failure(ex) => println(s"Failed to complete all futures: ${ex.getMessage}")
    }

  private def getSolutionFileName(marker : String,  solName : String, mapping : Map[String, String]) : String =
    // get solution name strip off the .md
    val solution = if mapping.contains(marker) then
      solName.replace(marker, mapping(marker))
    else
      solName

    solution.split(Utils.separatorRegEx).mkString("/")

  /**
   * given the solution name return the story name, if the story name is in mapping then use that rather
   * Note that separator is '/' and not the os separator. Vault path naming convention is '/'
   * @param solName to use for the story name
   * @return the story name
   */
  private def getStoryFileName(solName : SOLNAME, mapping : Map[String, String]) : String =
    val nameList = solName.split(Utils.separatorRegEx).drop(1)
    val storyName = nameList.last
    if mapping.contains(storyName) then
      s"${nameList.dropRight(1).mkString("/#")}${mapping(storyName)}"
    else
      nameList.mkString("/")

  /**
   * Take a marker string and convert into file path / file name.md. If the marker excluding the -[0-9]+
   * is in mapping then use that mapping value
   * @param marker in document string
   * @return
   */
  private def solutionDocNameFromMarker(solFolder : String, marker : String): SOLNAME =
    val markerList = marker.drop(1).split("-")
    val docName = markerList.dropRight(1).mkString("/")
    val solutionName = s"${docName}.md"
    s"""$solFolder${Utils.separator}$solutionName"""

  /**
   * Take a marker string and convert into file path
   *
   * @param marker
   * @return
   */
  private def solutionPathFromMarker(solFolder : String, marker: String): PATHNAME =
    val markerStr = marker.drop(1)
    val markerList = markerStr.split("-").dropRight(2)
    val pathName = s""""${markerList.dropRight(1).mkString("/")}"""

    s"""$solFolder${Utils.separator}$pathName"""

File: ts/CrossCuttingConcerns.ts
import { App, FileSystemAdapter } from 'obsidian';
import { Utils } from './Utils'
import { DocFolders } from './DocFolders';
import { statSync, readdirSync, mkdirSync } from 'fs'

export class CrossCuttingConcerns {
    private fsa: FileSystemAdapter;
    private docFolders: DocFolders;
    private utils: Utils;

    constructor(app: App, docFolders: DocFolders) {
        this.utils = new Utils(app);
        this.fsa = this.utils.fsa;
        this.docFolders = docFolders;
    }

    generateCrossCuttingConcerns() {
        const documentToMarkerMap = new Map<string, Set<string>>;
        const testDocumentToMarkerMap = new Map<string, Set<string>>;
        const storyToMarkerMap = new Map<string, Set<string>>;
        const testStoryToMarkerMap = new Map<string, Set<string>>;
        const markerToStoryMap = new Map<string, string>();
        const markerToTestStoryMap = new Map<string, string>();
        const markerToDocumentMap = new Map<string, string>();
        const markerToTestDocumentMap = new Map<string, string>();
        // get all the solutions files to delete
        //
        const promise: Array<Promise<void>> = new Array<Promise<void>>();
        const solutionFileToDelete = this.utils.listMDFilesInVault(this.docFolders.settingsSolutionFolder);
        solutionFileToDelete.forEach(file => {
            promise.push(this.fsa.remove(file));
        });
        //
        // wait for all promises to complete
        //
        Promise.allSettled(promise)
            .then(value => {
                //
                // cleanup target folders, by the time the folders are gotten 
                // in next lines this cleanup is done
                //
                this.fsa.rmdir(this.docFolders.settingsSolutionFolder, true);
                this.fsa.mkdir(this.docFolders.settingsSolutionFolder);
                //
                // get list of md files in the comment and test folder of the documents
                //
                const commentMDFiles = this.utils.listMDFilesInVault(this.docFolders.settingsCommentFolder);
                const testCommentMDFiles = this.utils.listMDFilesInVault(this.docFolders.settingsTestCommentFolder);
                //
                // get list of md files in the story folder
                //
                const storyMDFiles = this.utils.listMDFilesInVault(this.docFolders.settingsStoryFolder);
                const testStoryMDFiles = this.utils.listMDFilesInVault(this.docFolders.settingsUnitTestFolder);
                //
                // pick up all markers in the doc string doc file by doc file and aggregate the markers
                // before processing them
                //
                const allPromises = new Array<Promise<void>>();
                commentMDFiles.forEach(commentFile => {
                    allPromises.push(this.fsa.read(commentFile)
                        .then(value => {
                            var markerSet = new Set<string>();
                            //
                            // get all the markers in the value string
                            //
                            const markersMatch = value.matchAll(this.utils.markerRegExp);
                            Array.from(markersMatch).forEach(marker => {
                                markerSet.add(marker[0].trim());
                            })
                            markerSet = this.utils.sortSetOfString(markerSet);
                            const documentName = commentFile.split('/').last() as string;
                            markerSet.forEach(marker => {
                                markerToDocumentMap.set(marker.trim(), documentName);
                            });
                            if (documentName != undefined) {
                                documentToMarkerMap.set(documentName, markerSet)
                            }
                        }))
                })
                testCommentMDFiles.forEach(commentFile => {
                    allPromises.push(this.fsa.read(commentFile)
                        .then(value => {
                            var markerSet = new Set<string>();
                            //
                            // get all the markers in the value string
                            //
                            const markersMatch = value.matchAll(this.utils.markerRegExp);
                            Array.from(markersMatch).forEach(marker => {
                                markerSet.add(marker[0].trim());
                            })
                            markerSet = this.utils.sortSetOfString(markerSet);
                            const documentName = commentFile.split('/').last() as string;
                            markerSet.forEach(marker => {
                                markerToTestDocumentMap.set(marker.trim(), documentName);
                            });
                            if (documentName != undefined) {
                                testDocumentToMarkerMap.set(documentName, markerSet)
                            }
                        }))
                })
                storyMDFiles.forEach(storyFile => {
                    allPromises.push(this.fsa.read(storyFile)
                        .then(value => {
                            var markerSet = new Set<string>();
                            //
                            // get all the markers in the value string
                            //
                            const markersMatch = value.matchAll(this.utils.markerRegExp);
                            Array.from(markersMatch).forEach(marker => {
                                markerSet.add(marker[0].trim());
                            })
                            markerSet = this.utils.sortSetOfString(markerSet);
                            const documentName = storyFile.split('/').last() as string;
                            markerSet.forEach(marker => {
                                markerToStoryMap.set(marker.trim(), storyFile);
                            });
                            if (documentName != undefined) {
                                storyToMarkerMap.set(documentName, markerSet)
                            }
                        }))
                })
                testStoryMDFiles.forEach(storyFile => {
                    allPromises.push(this.fsa.read(storyFile)
                        .then(value => {
                            var markerSet = new Set<string>();
                            //
                            // get all the markers in the value string
                            //
                            const markersMatch = value.matchAll(this.utils.markerRegExp);
                            Array.from(markersMatch).forEach(marker => {
                                markerSet.add(marker[0].trim());
                            })
                            markerSet = this.utils.sortSetOfString(markerSet);
                            const documentName = storyFile.split('/').last() as string;
                            markerSet.forEach(marker => {
                                markerToTestStoryMap.set(this.dropRightAndMkString(marker.trim().split("-"), 4, "-"), storyFile);
                            });
                            if (documentName != undefined) {
                                testStoryToMarkerMap.set(documentName, markerSet)
                            }
                        }))
                })
                Promise.allSettled(allPromises)
                    .then(value => {
                        //
                        // collect all markers in one list
                        // sort them them
                        // group by path/name.md excluding the seq number
                        // 
                        var listOfMarkers: string[] = [];
                        var unitTestMarkers: string[] = [];

                        Array.from(documentToMarkerMap.values())
                            .forEach(setOfMarkers => {
                                const loalListOfMarkers: string[] = Array.from(setOfMarkers);
                                listOfMarkers = listOfMarkers.concat(loalListOfMarkers);
                            });
                        Array.from(testDocumentToMarkerMap.values())
                            .forEach(setOfMarkers => {
                                const loalListOfMarkers: string[] = Array.from(setOfMarkers);
                                unitTestMarkers = unitTestMarkers.concat(loalListOfMarkers);
                            });
                        listOfMarkers.sort((a, b) => a.localeCompare(b))
                        unitTestMarkers.sort((a, b) => a.localeCompare(b))
                        var allMarkers = this.groupedMap(listOfMarkers, i => this.solutionDocNameFromMarker(i));
                        var allUnitTestMarkers = this.groupedUnitTestMap(unitTestMarkers);
                        Array.from(allMarkers).forEach(([solName, markers]) => {
                            var mdString = `# ${this.dropRightAndMkString(solName.split("/"), 1, " ").toUpperCase()}\n`;
                            //
                            // build link to story, filter the story markers as well
                            //
                            var markerToStory: Map<string, string> = new Map(
                                Array.from(markerToStoryMap.entries())
                                    .filter(([key]) => {
                                        const splitMarker = key.split("-");
                                        return markers[0].startsWith(this.dropRightAndMkString(splitMarker, 1, '-'));
                                    }
                                    ));
                            //
                            // setup die story links first
                            //                                      
                            mdString = mdString + '## Functional Requirement\n';
                            markerToStory.forEach((story, marker) => {
                                if (markerToStoryMap.get(marker) != undefined) {
                                    mdString = mdString + `![[${markerToStoryMap.get(marker)}#${marker.trim()}]]\n`;
                                }
                            });

                            const uniqueMakers = Array.from(new Set(markers));
                            uniqueMakers.forEach((marker, story) => {
                                mdString = mdString + '## Implimentation Solution\n';
                                if (markerToDocumentMap.get(marker) != undefined) {
                                    const document = markerToDocumentMap.get(marker)!;
                                    mdString = mdString + `![[${document}#${marker.trim()}]]\n`;

                                    //
                                    // if the unit tests for the document exis then
                                    // print out these unit tests - there be dragons here
                                    //
                                    if (allUnitTestMarkers.get(marker) != undefined) {
                                        mdString = mdString + '### Unit Test Implementation\n';
                                        allUnitTestMarkers.get(marker)!.forEach(marker => {
                                            if (this.isTestMarker(marker)) {
                                                mdString = mdString + `![[${markerToTestDocumentMap.get(marker)}#${marker.trim()}]]\n`;
                                            }
                                        });
                                    }
                                }
                            });

                            //
                            // create the folder path if required and write out text
                            //
                            this.utils.makeDirInVault(solName);
                            this.fsa.write(solName, mdString);
                        })
                    })
            });
    }

    /**
     * 
     * @param values drop the rightmost nth string and join using the delimiter
     * @param n string to drop on the right
     * @param delimiter used to join 
     * @returns resulting string
     */
    private dropRightAndMkString(values: string[], n: number, delimiter: string): string {
        return values.slice(0, values.length - (n)).join(delimiter);
    }

    /**
     * 
     * @param values drop the leftmost nth string an return the result
     * @param n string to drop on the right
     * @param delimiter used to join 
     * @returns resulting string
     */
    private dropLeftAndMkString(values: string[], n: number, delimiter: string): string {
        return values.slice(n, values.length - (n)).join(delimiter);
    }

    /**
     * split solName in groups of 2 joined by /
     * @param solName solution name
     * @param mapping markerSet
     */
    private getSolutionFileName(solName: string): string {
        const fileNameParts = solName.split("-")
        var fileName: string[] = [];
        var i = 0;
        for (i = 0; i < fileNameParts.length; i++) {
            if (i % 2 == 0) {
                if (i == fileNameParts.length - 2) {
                    fileName.push(fileNameParts[i]);
                } else {
                    fileName.push(fileNameParts[i] + "-" + fileNameParts[i + 1]);
                }
            }
        }
        return `/${fileName.join("/")}`
    }

    /**
     * Take a marker string and convert into file path / file name.md. of the marker 
     * excluding the -[0-9]+ at the end, is in mapping then use that mapping value.
     * markers have the folllowing format :
     * 
     * 1. ^JIRA1234-001 <
     * 2. ^JIRA1234-001-solution-001
     * 3. ^JIRA1234-001-solution-001-test-001
     * 
     * 1 marker converted to filename
     * 2 marker converted to path + filename
     * 3 marker converted to path + filenanme where - identical to 2.
     * 
     * @param marker in document string
     * @return the name of the path and file name.md
     */
    private solutionDocNameFromMarker(marker: string): string {
        const docName = this.getSolutionFileName(marker.replace("^", ""));
        const solutionName = `${docName}.md`;
        return `/${this.docFolders.settingsSolutionFolder}${solutionName}`;
    }

    /**
     * Group by the list according to a getKey function. It will be one key 
     * to many potential values
     * @param list of elements to group by
     * @param getKey group by key
     * @returns the grouped list
     */
    private groupedMap(array: string[], getKey: (i: string) => string): Map<string, string[]> {
        return array.reduce((map, currentValue) => {
            const key = getKey(currentValue);

            if (!map.has(key)) {
                map.set(key, []);
            }

            if (map.get(key) != undefined) {
                map?.get(key)?.push(currentValue);
            }

            return map;
        }, new Map<string, string[]>());
    }

    /**
     * Group by the list according to a getKey function. It will be one key 
     * to many potential values
     * @param list of elements to group by, The list in this case will contain string formatted
     * as JIRA1234-001-solution-001-test-001. 
     * @param getKey group by key
     * @returns the grouped list. The Map<solName, Map<Document Marker marker link, List<string>(Unit test links)> >
     */
    private groupedUnitTestMap(array: string[]): Map<string, string[]> {
        return array.reduce((map, currentValue) => {
            const key = currentValue.split("-");
            const keyValue = this.dropRightAndMkString(key, 2, "-");

            if (!map.has(keyValue)) {
                map.set(keyValue, []);
            }

            if (map.get(keyValue) != undefined) {
                map?.get(keyValue)?.push(currentValue);
            }

            return map;
        }, new Map<string, string[]>());
    }

    /**
     * is this a test marker
     * @param marker to check is it has legth 6
     * @returns 
     */
    private isTestMarker(marker: string): boolean {
        return marker.split("-").length == 6
    }

}
